session/common.go

```diff
@@ -0,0 +1,334 @@
+// Calls contextReader if session context is set. Passes checkFieldPresence into contextReader.
+func (x *commonData) readFromV2(m session.Token, checkFieldPresence bool, r contextReader) error {
+	var err error
+
+	body := m.GetBody()
+	if checkFieldPresence && body == nil {
+		return errors.New("missing token body")
+	}
+
+	binID := body.GetID()
+	if x.idSet = len(binID) > 0; x.idSet {
+		err = x.id.UnmarshalBinary(binID)
+		if err != nil {
+			return fmt.Errorf("invalid session ID: %w", err)
+		} else if ver := x.id.Version(); ver != 4 {
+			return fmt.Errorf("invalid session UUID version %d", ver)
+		}
+	} else if checkFieldPresence {
+		return errors.New("missing session ID")
+	}
+
+	issuer := body.GetOwnerID()
+	if x.issuerSet = issuer != nil; x.issuerSet {
+		err = x.issuer.ReadFromV2(*issuer)
+		if err != nil {
+			return fmt.Errorf("invalid session issuer: %w", err)
+		}
+	} else if checkFieldPresence {
+		return errors.New("missing session issuer")
+	}
+
+	lifetime := body.GetLifetime()
+	if x.lifetimeSet = lifetime != nil; x.lifetimeSet {
+		x.iat = lifetime.GetIat()
+		x.nbf = lifetime.GetNbf()
+		x.exp = lifetime.GetExp()
+	} else if checkFieldPresence {
+		return errors.New("missing token lifetime")
+	}
+
+	x.authKey = body.GetSessionKey()
+	if checkFieldPresence && len(x.authKey) == 0 {
+		return errors.New("missing session public key")
+	}
+
+	c := body.GetContext()
+	if c != nil {
+		err = r(c, checkFieldPresence)
+		if err != nil {
+			return fmt.Errorf("invalid context: %w", err)
+		}
+	} else if checkFieldPresence {
+		return errors.New("missing session context")
+	}
+
+	sig := m.GetSignature()
+	if x.sigSet = sig != nil; sig != nil {
+		x.sig = *sig
+	} else if checkFieldPresence {
+		return errors.New("missing body signature")
+	}
+
+	return nil
+}
+
+type contextWriter func() session.TokenContext
+
+func (x commonData) fillBody(w contextWriter) *session.TokenBody {
+	var body session.TokenBody
+
+	if x.idSet {
+		binID, err := x.id.MarshalBinary()
+		if err != nil {
+			panic(fmt.Sprintf("unexpected error from UUID.MarshalBinary: %v", err))
+		}
+
+		body.SetID(binID)
+	}
+
+	if x.issuerSet {
+		var issuer refs.OwnerID
+		x.issuer.WriteToV2(&issuer)
+
+		body.SetOwnerID(&issuer)
+	}
+
+	if x.lifetimeSet {
+		var lifetime session.TokenLifetime
+		lifetime.SetIat(x.iat)
+		lifetime.SetNbf(x.nbf)
+		lifetime.SetExp(x.exp)
+
+		body.SetLifetime(&lifetime)
+	}
+
+	body.SetSessionKey(x.authKey)
+
+	body.SetContext(w())
+
+	return &body
+}
+
+func (x commonData) writeToV2(m *session.Token, w contextWriter) {
+	body := x.fillBody(w)
+
+	m.SetBody(body)
+
+	var sig *refs.Signature
+
+	if x.sigSet {
+		sig = &x.sig
+	}
+
+	m.SetSignature(sig)
+}
+
+func (x commonData) signedData(w contextWriter) []byte {
+	return x.fillBody(w).StableMarshal(nil)
+}
+
+func (x *commonData) sign(key ecdsa.PrivateKey, w contextWriter) error {
+	user.IDFromKey(&x.issuer, key.PublicKey)
+	x.issuerSet = true
```

Comment:
Can we do it after the `if err != nil`, so that `x` is unchanged in case of an error?

Reply:
Sure, done in b7efdf40f2367f59c3928bd3afbb30caddbad681